systemtest/benchtest/main.go

```diff
@@ -72,39 +78,47 @@ func runBenchmark(f BenchmarkFunc) (testing.BenchmarkResult, bool, error) {
 		limiter := getNewLimiter(maxEPM)
 		b.ResetTimer()
 		f(b, limiter)
-		for !b.Failed() {
-			if err := queryExpvar(&after, *server); err != nil {
+		if !b.Failed() {
+			watcher, err := collector.AddWatch(expvar.ActiveEvents, 0)
+			if err != nil {
 				b.Error(err)
-				break
-			}
-			if after.ActiveEvents == 0 {
-				break
+			} else if status := <-watcher; !status {
+				b.Error("failed to wait for APM server to be inactive")
 			}
-			time.Sleep(100 * time.Millisecond)
 		}
 		ok = !b.Failed()
 	})
 	if result.Extra != nil {
-		addExpvarMetrics(result, before, after)
+		addExpvarMetrics(result, collector, *detailed)
 	}
 	return result, ok, nil
 }
 
-func addExpvarMetrics(result testing.BenchmarkResult, before, after expvar) {
-	result.MemAllocs = after.MemStats.Mallocs - before.MemStats.Mallocs
-	result.MemBytes = after.MemStats.TotalAlloc - before.MemStats.TotalAlloc
-	result.Bytes = after.UncompressedBytes - before.UncompressedBytes
-	result.Extra["events/sec"] = float64(after.TotalEvents-before.TotalEvents) / result.T.Seconds()
-	result.Extra["txs/sec"] = float64(after.TransactionsProcessed-before.TransactionsProcessed) / result.T.Seconds()
-	result.Extra["spans/sec"] = float64(after.SpansProcessed-before.SpansProcessed) / result.T.Seconds()
-	result.Extra["metrics/sec"] = float64(after.MetricsProcessed-before.MetricsProcessed) / result.T.Seconds()
-	result.Extra["errors/sec"] = float64(after.ErrorsProcessed-before.ErrorsProcessed) / result.T.Seconds()
+func addExpvarMetrics(result testing.BenchmarkResult, collector *expvar.Collector, detailed bool) {
```

Comment:
This function is starting to contain ad decent amount of logic, could you write a few tests that cover the basic logic and the detailed flag?